include/swift/RemoteInspection/TypeRefBuilder.h

```diff
@@ -634,17 +635,58 @@ class TypeRefBuilder {
     if (!builtTypeDecl)
       return nullptr;
 
-    if (!builtTypeDecl->genericParamsPerLevel)
+    // If there aren't generic params on the parent types, we just emit
+    // a single BG typeref with all the generic args
+    auto maybeGenericParamsPerLevel = builtTypeDecl->genericParamsPerLevel;
+    if (!maybeGenericParamsPerLevel) {
       return BoundGenericTypeRef::create(*this, builtTypeDecl->mangledName, args, nullptr);
+    }
 
-  
+    // Otherwise, work from a full demangle tree to produce a
+    // typeref that includes information about parent generic args
     auto node = Dem.demangleType(builtTypeDecl->mangledName);
-    if (!node || !node->hasChildren() || node->getKind() != Node::Kind::Type)
+    if (!node || !node->hasChildren() || node->getKind() != Node::Kind::Type) {
+      return nullptr;
+    }
+    auto startNode = node->getFirstChild();
+    auto mangling = Demangle::mangleNode(startNode);
+    if (!mangling.isSuccess()) {
       return nullptr;
+    }
+
+    // Sanity:  Verify that the generic params per level add
+    // up exactly to the number of args we were provided, and
+    // that we don't have a rediculous number of either one
+    auto genericParamsPerLevel = *maybeGenericParamsPerLevel;
```

Comment:
There were a lot of checks to ensure that the shape list (params-per-level) and the actual args list made sense together.  I found it easier to pull that together into a single unified check:  Ensure that the total number of params mentioned in params-per-level adds up exactly to the number of args we have.  I also imposed an arbitrary limit of 1000 nested types and 1000 generic args in order to be comfortably certain that we'll never have problems from integer wrapping.